pkg/ring/kv/dynamodb/client.go

```diff
@@ -0,0 +1,313 @@
+}
+
+var (
+	backoffConfig = backoff.Config{
+		MinBackoff: 1 * time.Second,
+		MaxBackoff: 1 * time.Minute,
+		MaxRetries: 0,
+	}
+
+	defaultLoopDelay = 1 * time.Minute
+)
+
+// RegisterFlags adds the flags required to config this to the given FlagSet
+// If prefix is not an empty string it should end with a period.
+func (cfg *Config) RegisterFlags(f *flag.FlagSet, prefix string) {
+	f.StringVar(&cfg.Region, prefix+"dynamodb.region", "", "Region to access dynamodb.")
+	f.StringVar(&cfg.TableName, prefix+"dynamodb.table-name", "", "Table name to use on dynamodb.")
+	f.DurationVar(&cfg.TTL, prefix+"dynamodb.ttl-time", 0, "Time to expire items on dynamodb.")
+}
+
+func NewClient(cfg Config, cc codec.Codec, logger log.Logger, registerer prometheus.Registerer) (*Client, error) {
+	dynamoDB, err := newDynamodbKV(cfg, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	ddbMetrics := newDynamoDbMetrics(registerer)
+
+	c := &Client{
+		kv:         dynamodbInstrumentation{kv: dynamoDB, ddbMetrics: ddbMetrics},
+		codec:      cc,
+		logger:     logger,
+		ddbMetrics: ddbMetrics,
+		staleData:  make(map[string]staleData),
+	}
+	return c, nil
+}
+
+func (c *Client) List(ctx context.Context, key string) ([]string, error) {
+	resp, _, err := c.kv.List(ctx, dynamodbKey{primaryKey: key})
+	if err != nil {
+		level.Warn(c.logger).Log("msg", "error listing dynamodb", "key", key, "err", err)
+		return nil, err
+	}
+	return resp, err
+}
+
+func (c *Client) Get(ctx context.Context, key string) (interface{}, error) {
+	resp, _, err := c.kv.Query(ctx, dynamodbKey{primaryKey: key}, false)
+	if err != nil {
+		level.Warn(c.logger).Log("msg", "error Get dynamodb", "key", key, "err", err)
+		return nil, err
+	}
+	res, err := c.decodeMultikey(resp)
+	if err != nil {
+		return nil, err
+	}
+	c.updateStaleData(key, res)
+
+	return res, nil
+}
+
+func (c *Client) Delete(ctx context.Context, key string) error {
+	resp, _, err := c.kv.Query(ctx, dynamodbKey{primaryKey: key}, false)
+	if err != nil {
+		level.Warn(c.logger).Log("msg", "error Delete dynamodb", "key", key, "err", err)
+		return err
+	}
+
+	for innerKey := range resp {
+		err := c.kv.Delete(ctx, dynamodbKey{
+			primaryKey: key,
+			sortKey:    innerKey,
+		})
+		if err != nil {
+			level.Warn(c.logger).Log("msg", "error Delete dynamodb", "key", key, "innerKey", innerKey, "err", err)
+			return err
+		}
+	}
+	return err
+}
+
+func (c *Client) CAS(ctx context.Context, key string, f func(in interface{}) (out interface{}, retry bool, err error)) error {
+	bo := backoff.New(ctx, backoffConfig)
+	for bo.Ongoing() {
+		resp, _, err := c.kv.Query(ctx, dynamodbKey{primaryKey: key}, false)
+		if err != nil {
+			level.Error(c.logger).Log("msg", "error cas query", "key", key, "err", err)
+			bo.Wait()
+			continue
+		}
+
+		current, err := c.decodeMultikey(resp)
+		if err != nil {
+			level.Error(c.logger).Log("msg", "error decoding key", "key", key, "err", err)
+			continue
+		}
+
+		out, retry, err := f(current.Clone())
```

Comment:
Will `f` change current here? Why we want to clone?

Reply:
Yes, f does change the data of passed obj. We need the original obj to get difference of the changes with `FindDifference`